File: netcore/src/Koralium.SqlParser/QueryBuilder/QueryBuilder.cs
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using Koralium.SqlParser.Expressions;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;

namespace Koralium.SqlParser
{
    /// <summary>
    /// Class that helps building up specific parts of a query using expressions
    /// </summary>
    /// <typeparam name="TEntity"></typeparam>
    public static class QueryBuilder
    {
        /// <summary>
        /// Create a boolean sql expression using expressions
        /// </summary>
        /// <param name="expression"></param>
        /// <returns></returns>
        public static BooleanExpression BooleanExpression<TEntity>(Expression<Func<TEntity, bool>> expression)
        {
            var evaluatedExpression = PartialEvaluator.PartialEval(expression) as System.Linq.Expressions.LambdaExpression;
            var visitor = new FilterExpressionVisitor();

            visitor.Visit(evaluatedExpression.Body);
            return visitor.BooleanExpression;
        }

        public static BooleanExpression And(params BooleanExpression[] booleanBinaryExpressions)
        {
            return Binary_Internal(booleanBinaryExpressions, BooleanBinaryType.AND);
        }

        public static BooleanExpression And(IEnumerable<BooleanExpression> expressions)
        {
            return Binary_Internal(expressions, BooleanBinaryType.AND);
        }

        private static BooleanExpression Binary_Internal(IEnumerable<BooleanExpression> expressions, BooleanBinaryType booleanBinaryType)
        {
            if(!(expressions is IReadOnlyList<BooleanExpression> list))
            {
                list = expressions.ToList();
            }
            if(list.Count == 0)
            {
                return null;
            }
            var first = list.First();

            if(list.Count == 1)
            {
                return first;
            }

            var second = list[1];

            var binaryExpression = new BooleanBinaryExpression()
            {
                Left = first,
                Right = second,
                Type = booleanBinaryType
            };

            for(int i = 2; i < list.Count; i++)
            {
                binaryExpression = new BooleanBinaryExpression()
                {
                    Left = binaryExpression,
                    Right = list[i],
                    Type = booleanBinaryType
                };
            }

            return binaryExpression;
        }

        public static BooleanExpression Or(params BooleanExpression[] expressions)
        {
            return Binary_Internal(expressions, BooleanBinaryType.OR);
        }

        public static BooleanExpression Or(IEnumerable<BooleanExpression> expressions)
        {
            return Binary_Internal(expressions, BooleanBinaryType.OR);
        }
    }
}

File: presto/trino-connector-arrowflight/src/main/java/io/trino/plugin/koralium/client/QueryBuilder.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.trino.plugin.koralium.client;

import io.trino.plugin.koralium.KoraliumPrestoColumn;
import io.trino.plugin.koralium.KoraliumSortItem;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.OptionalLong;
import java.util.stream.Collectors;

public class QueryBuilder
{
    public String buildCountQuery(
            String tableName,
            String filter,
            OptionalLong limit)
    {
        if (limit.isPresent()) {
            StringBuilder stringBuilder = new StringBuilder();
            stringBuilder.append("SELECT count(*) FROM ( SELECT * FROM ");
            stringBuilder.append(tableName);

            if (filter != null && !filter.isEmpty()) {
                stringBuilder.append(" WHERE ");
                stringBuilder.append(filter);
            }

            stringBuilder.append(" LIMIT ");
            stringBuilder.append(limit.getAsLong());

            stringBuilder.append(")");
            return stringBuilder.toString();
        }
        else {
            StringBuilder stringBuilder = new StringBuilder();
            stringBuilder.append("SELECT count(*) FROM ");
            stringBuilder.append(tableName);

            if (filter != null && !filter.isEmpty()) {
                stringBuilder.append(" WHERE ");
                stringBuilder.append(filter);
            }

            return stringBuilder.toString();
        }
    }

    public String buildQuery(
            Optional<List<KoraliumPrestoColumn>> columns,
            String tableName,
            String filter,
            Optional<List<KoraliumSortItem>> sortItems,
            OptionalLong limit)
    {
        StringBuilder stringBuilder = new StringBuilder();

        stringBuilder.append("SELECT ");

        if (columns.isPresent()) {
            String columnSelect = String.join(", ", columns.get().stream().map(x -> x.getColumnName()).collect(Collectors.toList()));
            stringBuilder.append(columnSelect);
        }
        else {
            stringBuilder.append("*");
        }

        stringBuilder.append(" FROM ");
        stringBuilder.append(tableName);

        if (filter != null && !filter.isEmpty()) {
            stringBuilder.append(" WHERE ");
            stringBuilder.append(filter);
        }

        if (sortItems.isPresent()) {
            List<String> orderBys = new ArrayList<>();
            List<KoraliumSortItem> items = sortItems.get();

            for (KoraliumSortItem item : items) {
                String sortString = item.getColumnHandle().getColumnName();

                if (item.getSortOrder().isAscending()) {
                    sortString += " ASC";
                }
                else {
                    sortString += " DESC";
                }
                orderBys.add(sortString);
            }

            if (orderBys.size() > 0) {
                stringBuilder.append(" ORDER BY ");
                stringBuilder.append(String.join(", ", orderBys));
            }
        }

        if (limit.isPresent()) {
            stringBuilder.append(" LIMIT ");
            stringBuilder.append(limit.getAsLong());
        }

        return stringBuilder.toString();
    }
}
